### ISSUE TITLE
Issues with using model bundle for prediction

### ISSUE THREAD

**Issue by xholmes** [azavea/raster-vision#1876]
Hi, 

Thank you for the awesome work on developing raster-vision. 
I had manage to trained a model for semantic segmentation on satellite images using code adapted from [the ISPRS Potsdam example](https://github.com/azavea/raster-vision/blob/0.20/rastervision_pytorch_backend/rastervision/pytorch_backend/examples/semantic_segmentation/isprs_potsdam.py).

However, I had run into issues using the model bundle for prediction. I had tried several ways and had ran into either failures or a varying degree of success. I have listed each of my approach and level of success below. Would very much appreciate your views on how I can use the model bundle effectively for prediction on unseen images/data.

Regardless of approach, my goal is to predict on a list of unseen images and to generate a prediction tif and vector file for each class.

Approach 1 - Using `Predictor` class

For the 1st approach, I had tried to use the `Predictor` class. I had tried this method with RV v0.12 with success. However, come v0.20, I got the following error:

```
Traceback (most recent call last):
  File ".../v020_scripts/predict_pipeline.py", line 31, in <module>
    predictor = Predictor(args.m, args.t)
  File ".../venv/lib/python3.9/site-packages/rastervision/core/predictor.py", line 58, in __init__
    self.scene: 'SceneConfig' = self.config.dataset.validation_scenes[0]
AttributeError: 'LearnerPipelineConfig' object has no attribute 'dataset'
```
Note: I redacted some of the path information so the path looks weird. Please ignore the weird file paths.

The bundle I used was the `model-bundle.zip` file in the `bundle` folder, generated by the training script. I have also tried using the model bundle file found in the `train` folder, but that too, produced the same error.

For reference, key lines of the code used for this approach is as follows:

```
    # Specify the model bundle file and tmp directory 
    predictor = Predictor(args.m, args.t)
    print("Model loaded in {}".format(time.perf_counter() - t_start))

    label_format = '{}_label.tif'
    vector_format = '{}_{}_vector.json'
    predict_tif = os.path.join(args.o, 'labels.tif')
    predict_vector_path = os.path.join(args.o, 'predict')

    # Run predict on one image at a time
    o_start = time.perf_counter()
    for img in img_list:
        t_start = time.perf_counter()
        predictor.predict([img], args.o)
```

Approach 2: Using the `predict` command-line command

The 2nd approach I tried was to use the `predict` command-line command. For this approach, I used the `model-bundle.zip` file found in the `bundle` folder generated when I finish the model training. However, I did not get far with this approach as I kept getting the following error:

```
Error: Missing argument 'MODEL_BUNDLE`
```

This is despite having provided the module bundle in the following format:
1) specifying the exact path to the `model-bundle.zip` file
2) specifying the folder rather than the file
3) unzip the `model-bundle.zip` file and specify the folder
4) unzip the `model-bundle.zip` file and specify the `model-bundle.zip` file (there was another model-bundle.zip file  in the zip file)

Approach 3: 
My first approach was to use the `Learner` class as suggested [here](https://docs.rastervision.io/en/0.20/usage/tutorials/pred_and_eval_ss.html#Load-a-Learner-with-a-trained-model-from-bundle----Learner.from_model_bundle()). Following is the script put together:

```
    chip_sz = 64
    stride = math.floor(chip_sz / 4)

    label_classes = ['background', 'imprevious', 'building', 'tree', 'bare_earth']
    # black, yellow, blue, green, orange
    label_colours = ['#000000', '#f2e70c', '#0000ff', '#36ff00', '#e87400']

    class_config = ClassConfig(names=label_classes, colors=label_colours,
                               null_class='background')

    # Load data to use for inference
    img_set = [f for f in os.listdir(args.i) if f.endswith('.tif')]

    # Load pre-trained model for inference
    learner = SemanticSegmentationLearner.from_model_bundle(
        args.b, training=False
    )
    vector_folder = os.path.join(args.o, 'vector')
    vector_uri = os.path.join(args.o, 'vector', '{}_class{}.json')
    os.makedirs(vector_folder, exist_ok=True)

    for f in img_set:
        img = os.path.join(args.i, f)
        fname = f.split('.')[0]
        output_uri = os.path.join(args.o, fname)
        os.makedirs(output_uri)

        raster_source = RasterioSource(img)
        stats_transformer = StatsTransformer.from_raster_sources(
            raster_sources=[raster_source],
            max_stds=3
        )

        ds = SemanticSegmentationSlidingWindowGeoDataset.from_uris(
            class_config=class_config,
            image_uri=img,
            size=chip_sz,
            stride=stride,
            image_raster_source_kw={
                'channel_order': [0, 1, 2],
                'raster_transformers': [stats_transformer]
            }
        )

        predictions = learner.predict_dataset(
            ds,
            raw_out=True,
            numpy_out=True,
            predict_kw=dict(out_shape=(chip_sz, chip_sz)),
            progress_bar=True
        )

        # Predict on image and save labels as vectors
        pred_labels = SemanticSegmentationLabels.from_predictions(
            ds.windows,
            predictions,
            smooth=True,
            extent=ds.scene.extent,
            num_classes=len(label_classes)
        )

        pred_labels.save(
            uri=output_uri,
            crs_transformer=ds.scene.raster_source.crs_transformer,
            class_config=class_config,
            vector_outputs=[
                PolygonVectorOutputConfig(
                    class_id=1, uri=vector_uri.format(fname, 1)
                ),
                PolygonVectorOutputConfig(
                    class_id=2, uri=vector_uri.format(fname, 2)
                ),
                PolygonVectorOutputConfig(
                    class_id=3, uri=vector_uri.format(fname, 3)
                ),
                PolygonVectorOutputConfig(
                    class_id=4, uri=vector_uri.format(fname, 4)
                )
            ]
        )
```

I have tried this approach with the bundle found in the `train` folder. With this approach, I can successfully run the prediction. However, the prediction results is much poorer compared to the trained results (as indicated in the validation images used during training).

## Environment

 - Installed via pip install, ran on command line
 - Raster Vision version or commit: RV v0.20
 - OS (e.g., Linux): Linux
 - Python version: 3.9.5
 - CUDA 11.6.2

**Comment by AdeelH** [comment]
Hi, thank you for the detailed report and sorry that you're having problems with RV.

## Approach 1

I believe this error implies you are using the model bundle from the `train/` folder instead of the one in the `bundle/` folder, which is what you are supposed to use here.

## Approach 2
I think I know what is going on here. You are probably specifying `--channel-order` before the arguments to the `predict` command, something like `rastervision predict --channel-order 0 1 2 <model_bundle_uri> <image_uri> <out_uri>` and this is running into [this hacky bit of code](https://github.com/azavea/raster-vision/blob/0.20/rastervision_core/rastervision/core/cli.py#L9) that parses the channel indices given to `--channel-order`.

The workaround is to simply specify `--channel-order` at the end of the command. Like so:

```sh
rastervision predict \
https://s3.amazonaws.com/azavea-research-public-data/raster-vision/examples/model-zoo-0.20/isprs-potsdam-ss/model-bundle.zip \
https://s3.amazonaws.com/azavea-research-public-data/raster-vision/examples/model-zoo-0.20/isprs-potsdam-ss/sample-predictions/sample-img-isprs-potsdam-ss.tif \
prediction_output/ \
--channel-order 0 1 2
```

## Approach 3

One reason for the worse results could be that you are using different stats to normalize the images with the `StatsTransformer` than the ones you used during training. You can try using the stats from `analyze/` with your new data.

It is also possible that your new data is significantly different from your training set, in which case, you would see worse results if your model does not generalize well. So I would recommend predicting on images from your training/validation set as a sanity check.

**Comment by xholmes** [comment]
Hi @AdeelH,

Thank you for the prompt reply.

For approach 1, I had tried using both bundles from the `bundle` folder and the `train` folder. Both gives the same error.

Will give your suggestion on the order of the parameters a go for approach 2.

As for approach 3, I also thought of using the same validation image for a sanity check. The following is a comparison of the output of the validation during training vs output of the prediction (same image). This was done using the same model.  (please ignore the difference in colour as I did not play around with the colour display of the outputs, but you can notice the lack of class predictions in the predicted output from the large regions of blacks (background)).

Output of validation (during training)
<img width="412" alt="Screenshot 2023-09-01 at 11 27 22 AM" src="https://github.com/azavea/raster-vision/assets/6213484/c97242d6-a06b-48ca-afbc-6ebbf00495e6">

Output of prediction
<img width="412" alt="Screenshot 2023-09-01 at 11 25 50 AM" src="https://github.com/azavea/raster-vision/assets/6213484/e83f9721-8efa-4d49-84ea-a4db5659f56a">

**Comment by xholmes** [comment]
I gave the change of `--channel-order 0 1 2` to the end of the command-line a go. As you suggested, that solved the issue of missing bundle error. However, it ran into the same missing attribute error:

```
AttributeError: 'LearnerPipelineConfig' object has no attribute 'dataset'
```

I was using the bundle in the `bundle` folder. I suspect it might be the same issue as the one faced in Approach 1.

**Comment by AdeelH** [comment]
> I gave the change of `--channel-order 0 1 2` to the end of the command-line a go. As you suggested, that solved the issue of missing bundle error. However, it ran into the same missing attribute error:
> 
> ```
> AttributeError: 'LearnerPipelineConfig' object has no attribute 'dataset'
> ```

Interesting. That error is precisely what happens when you pass `train/model-bundle.zip` to the `Predictor` instead of `bundle/model-bundle.zip`. Can you confirm that the two bundles are not exactly the same? Are `pipeline-config.json`s in each of them a different size than the other? And does `bundle/model-bundle.zip` have another `model-bundle.zip` inside of it (i.e. inside the zip archive)?

I would recommend doing a fresh run (you can just run for 1 epoch to make it go faster) and then trying with the new bundle.

Also, are you able to successfully run the example predict command in my previous comment?

---

As for approach 1, can you try visualizing the predicted scores ([like in the tutorial](https://docs.rastervision.io/en/0.20/usage/tutorials/pred_and_eval_ss.html#Visualize-predictions)) to check if they are messed up before they are saved to file?

**Comment by xholmes** [comment]
> Interesting. That error is precisely what happens when you pass train/model-bundle.zip to the Predictor instead of bundle/model-bundle.zip. Can you confirm that the two bundles are not exactly the same? Are pipeline-config.jsons in each of them a different size than the other? And does bundle/model-bundle.zip have another model-bundle.zip inside of it (i.e. inside the zip archive)?

Right! That might be where I mucked up. Upon checking the zip file in the `bundle/` folder, it seems I had earlier unzipped the file. I have replaced the original file and am rerunning it now. Will update when I get the results back. 

But to answer your previous question, following is the comparison when I unzip both bundles:
Content of `bundle/model-bundle.zip`
![Screenshot 2023-09-04 at 11 57 48 AM](https://github.com/azavea/raster-vision/assets/6213484/3a747697-be20-41e7-9488-af90a1243c31)

Content of `train/model-bundle.zip`
![Screenshot 2023-09-04 at 11 57 37 AM](https://github.com/azavea/raster-vision/assets/6213484/9dc33dd8-7378-4b97-bfac-459fb3975217)

<hr/>

> Also, are you able to successfully run the example predict command in my previous comment?

After changing the order of the `-channel-order` flag, the command worked. However, I ran into the `no attribute 'dataset'` error. As you pointed out above, it is likely I went to unzip the `bundle/model-bundle.zip` file previously as I was exploring the file. I have re-run the job now. Waiting for the results to come back. Will update when I get back the results.

**Comment by xholmes** [comment]
Hi @AdeelH,

So after following your suggestions, I have managed to get all 3 approaches running successfully! 

For `approach 1`, where I was using the `Predictor` class with the `bundle/model-bundle.zip`, the main issue was that I had previously unzipped the zip file in the bundle folder and forgotten about it. Replacing it with the original bundle file worked out well. Outputs of the prediction were also more in-line with the performance of the trained model. 
![Screenshot 2023-09-05 at 11 21 45 AM](https://github.com/azavea/raster-vision/assets/6213484/bb896ecf-636a-45c3-932c-866809121c36)

<hr/>

With `approach 2`, where I was using the `rastervision predict` on the command-line, the re-positioning of the `--channel-order` flag to the end worked like a charm. The rest was due to me using the unzipped `bundle/model-bundle.zip`. Prediction results were the same as the one above (also because I was using the same model file).

<hr/>

Still waiting on the results from using the `Learner` class from Approach 3. Cluster had been quite busy these few days. I had removed the use of `StatsTransformer` from the pipeline. However, I suspect it won't do much good. The difference in the prediction is too large. Will update again when I get the results.

**Comment by xholmes** [comment]
Did as you suggested to check the prediction results when using the `Learner` class. I used the `train/model-bundle.zip` for reference. The outcome of the prediction seems to still vary significantly from the model's performance after training (see the post above for the expected outcome of the trained model).

Visual inspection of the prediction output saved as vectors.
![Screenshot 2023-09-05 at 5 45 59 PM](https://github.com/azavea/raster-vision/assets/6213484/be306616-c519-4c91-a7d1-448c4d01984c)

Prediction score before output is being saved

![ham_may_test1_scores](https://github.com/azavea/raster-vision/assets/6213484/8a2d662b-f50b-4fe9-8f48-7091f7f750cc)

**Comment by AdeelH** [comment]
Awesome! Glad you were able to get the `Predictor` working. 

If you have ruled out `StatsTransformer` as a problem in approach 3, my next guess would be a discrepancy in either the `chip_sz` or the `img_sz`. Ideally, they should be the same values as used for training. Note that for `img_sz` (which, by the way, is the size that chips are resized to before being passed to the model), you will need to explicitly pass it as a resize transform (`transform=A.Resize(img_sz, img_sz)`) to `SemanticSegmentationSlidingWindowGeoDataset` as shown here: https://docs.rastervision.io/en/stable/usage/tutorials/pred_and_eval_ss.html#Get-scene-to-predict.

**Comment by xholmes** [comment]
> Note that for img_sz (which, by the way, is the size that chips are resized to before being passed to the model)

I actually did not try to resize the chips, both during training and prediction. Instead, I was thinking to crop/chip them into smaller chunks of size 64. I did this during training as follows:

```
data_config = SemanticSegmentationGeoDataConfig(
            scene_dataset=scene_dataset,
            window_opts=GeoDataWindowConfig(method=GeoDataWindowMethod.sliding,
                                            size=chip_sz, stride=stride),
            img_channels=len(channel_order)
```
where `chip_sz` is 64 and `stride` is 1/4 of `chip_sz` (I wanted some overlaps). However, revisiting my code (I did this code quite sometime back), I am wondering if what I did was correct as I noticed that I had also declared a `SemanticSegmentationChipOptions` instance 

```
    chip_options = SemanticSegmentationChipOptions(
        window_method=SemanticSegmentationWindowMethod.sliding, stride=chip_sz
    )
```
and gave it to the Config
```
pipeline = SemanticSegmentationConfig(
        root_uri=output_uri,
        dataset=scene_dataset,
        backend=backend_config,
        train_chip_sz=chip_sz,
        predict_chip_sz=chip_sz,
        chip_options=chip_options
    )
```
Apologies, I digressed. As I did not do any resizing, in my prediction script in `approach 3`, I specified the chip sizes as follows:
```
        ds = SemanticSegmentationSlidingWindowGeoDataset.from_uris(
            class_config=class_config,
            image_uri=img,
            size=chip_sz,
            stride=stride,
            image_raster_source_kw={
                'channel_order': [0, 1, 2]
            }
        )

        predictions = learner.predict_dataset(
            ds,
            raw_out=True,
            numpy_out=True,
            predict_kw=dict(out_shape=(chip_sz, chip_sz)),
            progress_bar=True
        )
```

As a sanity check, can you help verify if what I had done does what I had intended?

**Comment by AdeelH** [comment]
Ah. `img_sz` is a field in `GeoDataConfig` which is 256 by default. See [the docs here for SemanticSegmentationGeoDataConfig.img_sz](https://docs.rastervision.io/en/0.20/api_reference/_generated/rastervision.pytorch_learner.semantic_segmentation_learner_config.SemanticSegmentationGeoDataConfig.html#rastervision.pytorch_learner.semantic_segmentation_learner_config.SemanticSegmentationGeoDataConfig.img_sz). So if you did not explicitly set it, 256 was the value used. So your model expects its inputs to be 256x256.

If you add the resize transform in my previous comment, your approach should hopefully work better. A more automatic way to do this would be to do something like:
```py
tf, _ = learner.cfg.data.get_data_transforms()
...
ds = SemanticSegmentationSlidingWindowGeoDataset.from_uris(
    ...,
    transform=tf
 )
```

`DataConfig.get_data_transforms()` returns a tuple of transforms, the first one of which is a "base" transform that is applied to both training and validation datasets and the second one is an augmentation transform that is applied to only the training dataset.

This is basically [what Raster Vision does internally](https://github.com/azavea/raster-vision/blob/0.20/rastervision_pytorch_backend/rastervision/pytorch_backend/pytorch_semantic_segmentation.py#L80-L87) when you predict via the `Predictor`.

>As a sanity check, can you help verify if what I had done does what I had intended?

Your use of `chip_sz` and `stride` is correct.

**Comment by xholmes** [comment]
> So if you did not explicitly set it, 256 was the value used. So your model expects its inputs to be 256x256.

Omg! Now everything clicks. That probably explains why I see artifacts like the the ones below in my predicted outputs (from using the `Predictor`).
![Screenshot 2023-09-06 at 1 25 17 PM](https://github.com/azavea/raster-vision/assets/6213484/b5dc7958-4cae-421a-a146-f257a8f07c46)

Ok, I will retrain my models but this time specifying `img_sz`. Thank you so much for the insights @AdeelH !

**Comment by xholmes** [comment]
> If you add the resize transform in my previous comment, your approach should hopefully work better. A more automatic way to do this would be to do something like:

Tried this and `approach 3` using the `Learner` class worked now.  Setting `smooth` to `True` in `SemanticSegmentationLabels` also helped with the boundary artifacts (I think).
![Screenshot 2023-09-07 at 4 19 03 PM](https://github.com/azavea/raster-vision/assets/6213484/881c3472-126c-40a5-9295-1f6b1ee4f5bf)

Thank you @AdeelH for your help! Now everything is working. Quick question though, what is the difference between using `SemanticSegmentationLabels.from_predictions` with `smooth=True` and using `SemanticSegmentationSmoothLabels.from_predictions`?

**Comment by AdeelH** [comment]
>Thank you @AdeelH for your help! Now everything is working.

Wonderful!



> Quick question though, what is the difference between using `SemanticSegmentationLabels.from_predictions` with `smooth=True` and using `SemanticSegmentationSmoothLabels.from_predictions`?

They are equivalent.

>Setting smooth to True in `SemanticSegmentationLabels` also helped with the boundary artifacts (I think).

It should, yes. Using `stride` < `chip_sz` during prediction also makes it so that (for non-corner pixels) the same pixel gets predicted multiple times (once for each chip that it is a part of) and all of its predictions averaged. It, however, also makes prediction slower, so you should try to find a `stride` that strikes good balance between speed and quality. An additional way to deal with boundary artifacts is the [`crop_sz` argument](https://docs.rastervision.io/en/0.20/api_reference/_generated/rastervision.core.data.label.semantic_segmentation_labels.SemanticSegmentationLabels.html#rastervision.core.data.label.semantic_segmentation_labels.SemanticSegmentationLabels.from_predictions); see an example [here](https://www.azavea.com/blog/2023/01/05/introducing-raster-vision-v0-20/#cleaner_semantic_segmentation_output).

**Comment by xholmes** [comment]
> It should, yes. Using stride < chip_sz during prediction also makes it so that (for non-corner pixels) the same pixel gets predicted multiple times (once for each chip that it is a part of) and all of its predictions averaged. It, however, also makes prediction slower, so you should try to find a stride that strikes good balance between speed and quality. An additional way to deal with boundary artifacts is the [crop_sz argument](https://docs.rastervision.io/en/0.20/api_reference/_generated/rastervision.core.data.label.semantic_segmentation_labels.SemanticSegmentationLabels.html#rastervision.core.data.label.semantic_segmentation_labels.SemanticSegmentationLabels.from_predictions); see an example [here](https://www.azavea.com/blog/2023/01/05/introducing-raster-vision-v0-20/#cleaner_semantic_segmentation_output).

Yes, while searching for how to tackle the boundary artifacts, I came across that post of yours. Will play around with the different parameters.

Once again, thank you for your help. Much appreciated. I'll close the issue since everything's working now.
